source/common/http/filter/gzip_filter.cc

```diff
@@ -0,0 +1,186 @@
+  if (!end_stream && !skip_compression_ && isMinimumContentLength(headers) &&
+      isContentTypeAllowed(headers) && isCacheControlAllowed(headers) && isEtagAllowed(headers) &&
+      isLastModifiedAllowed(headers) && isTransferEncodingAllowed(headers) &&
+      !headers.ContentEncoding()) {
+    headers.removeContentLength();
+    headers.insertContentEncoding().value(Http::Headers::get().ContentEncodingValues.Gzip);
+    compressor_.init(config_->compressionLevel(), config_->compressionStrategy(),
+                     config_->windowBits(), config_->memoryLevel());
+  } else {
+    skip_compression_ = true;
+  }
+
+  return Http::FilterHeadersStatus::Continue;
+}
+
+FilterDataStatus GzipFilter::encodeData(Buffer::Instance& data, bool end_stream) {
+  if (skip_compression_) {
+    return Http::FilterDataStatus::Continue;
+  }
+
+  const uint64_t n_data{data.length()};
+
+  if (n_data) {
+    compressor_.compress(data, compressed_data_);
+  }
+
+  if (end_stream) {
+    compressor_.flush(compressed_data_);
+  }
+
+  if (compressed_data_.length()) {
+    data.drain(n_data);
+    data.move(compressed_data_);
+    return Http::FilterDataStatus::Continue;
+  }
+
+  return Http::FilterDataStatus::StopIterationNoBuffer;
+}
+
+bool GzipFilter::isMinimumContentLength(HeaderMap& headers) const {
+  if (headers.ContentLength()) {
+    uint64_t content_length{};
+    StringUtil::atoul(headers.ContentLength()->value().c_str(), content_length);
+    return content_length >= config_->minimumLength();
+  }
+
+  return headers.TransferEncoding() &&
+         headers.TransferEncoding()->value().find(
+             Http::Headers::get().TransferEncodingValues.Chunked.c_str());
+}
+
+bool GzipFilter::isContentTypeAllowed(HeaderMap& headers) const {
+  if (config_->contentTypeValues().size() && headers.ContentType()) {
+    std::cmatch match;
+    std::regex_search(headers.ContentType()->value().c_str(), match, contentTypeRegex());
+    return config_->contentTypeValues().find(match.str()) != config_->contentTypeValues().end();
+  }
+  return true;
+}
+
+bool GzipFilter::isCacheControlAllowed(HeaderMap& headers) const {
+  if (config_->cacheControlValues().size() && headers.CacheControl()) {
+    return std::any_of(config_->cacheControlValues().begin(), config_->cacheControlValues().end(),
+                       [&headers](const auto& value) {
+                         return headers.CacheControl()->value().find(value.c_str());
+                       });
+  }
+  return true;
+}
+
+bool GzipFilter::isEtagAllowed(HeaderMap& headers) const {
+  if (headers.Etag()) {
+    return !config_->disableOnEtag();
+  }
+  return true;
+}
+
+bool GzipFilter::isLastModifiedAllowed(HeaderMap& headers) const {
+  if (headers.LastModified()) {
+    return !config_->disableOnLastModified();
+  }
+  return true;
+}
+
+bool GzipFilter::isTransferEncodingAllowed(HeaderMap& headers) const {
+  if (headers.TransferEncoding()) {
```

Comment:
probably worth doing a pass over all these methods and storing header lookups in temps rather than doing them twice.

Reply:
There is a great chance that the second call to TransferEncoding() might never happen, reason why this is also the second last checking in the if statement. Do you think that it's worth doing all these assignments to temps?

Reply:
IMO yes...HeaderMap's impl currently instantiates a discrete member variable per common header so the lookup is fast, but I'm wondering if that might be something that gets re-tuned later as that policy might be a drag on header construction speed as the number of potentially interesting headers grows.  And in any case the temp is free, so I don't see a downside to a temp in readability or performance.